### PR TITLE
Dyn stream reassembly v16

### PR DESCRIPTION
**Pull request by regit**
Update of #2307 addressing remarks and rebased.

PR builds:
- PR regit: https://buildbot.openinfosecfoundation.org/builders/regit/builds/212
- PR regit-pcap: https://buildbot.openinfosecfoundation.org/builders/regit-pcap/builds/208

### CODE REVIEW

File: doc/userguide/configuration/suricata-yaml.rst [hunk 1]
```diff
@@ -2050,3 +2050,21 @@ in which two threads have to wait for each other . When using two
...
+is recommended to keep the TCP connection opened with a remote device
+and not to open and close it for each MODBUS/TCP transaction.
+In that case, it is important to set the stream-depth of the modbus as
+unlimited.
+
+::
+
+      modbus:
+      # Stream reassembly size for modbus, default is 0
+      stream-depth: 0
```
Comment (inliniac) [inline doc/userguide/configuration/suricata-yaml.rst]: formatting is off, needs indent

File: src/app-layer-modbus.c [hunk 2]
```diff
@@ -3107,6 +3127,109 @@ static int ModbusParserTest16(void) {
...
+
+    FLOWLOCK_WRLOCK(&f);
+    r = AppLayerParserParse(NULL, alp_tctx, &f, ALPROTO_MODBUS, STREAM_TOCLIENT,
+                            readCoilsRsp, sizeof(readCoilsRsp));
+    FAIL_IF(r != 0);
+    FLOWLOCK_UNLOCK(&f);
+
+    FAIL_IF(((TcpSession *)(f.protoctx))->reassembly_depth != MODBUS_CONFIG_DEFAULT_STREAM_DEPTH);
+
+    if (alp_tctx != NULL)
```
Comment (inliniac) [inline src/app-layer-modbus.c]: we shouldn't have conditions in tests

File: src/app-layer-modbus.c [hunk 3]
```diff
@@ -3107,6 +3127,109 @@ static int ModbusParserTest16(void) {
...
+
+    FLOWLOCK_WRLOCK(&f);
+    r = AppLayerParserParse(NULL, alp_tctx, &f, ALPROTO_MODBUS, STREAM_TOCLIENT,
+                            input, input_len);
+    FAIL_IF(r != 0);
+    FLOWLOCK_UNLOCK(&f);
+
+    FAIL_IF(((TcpSession *)(f.protoctx))->reassembly_depth != MODBUS_CONFIG_DEFAULT_STREAM_DEPTH);
+
+    if (alp_tctx != NULL)
```
Comment (inliniac) [inline src/app-layer-modbus.c]: cond

File: src/app-layer-parser.h [hunk 4]
```diff
@@ -149,6 +150,9 @@ void AppLayerParserRegisterDetectStateFuncs(uint8_t ipproto, AppProto alproto,
         int (*StateHasTxDetectState)(void *alstate),
         DetectEngineState *(*GetTxDetectState)(void *tx),
         int (*SetTxDetectState)(void *alstate, void *tx, DetectEngineState *));
+void AppLayerParserRegisterGetStreamDepth(uint8_t ipproto,
+                                          AppProto alproto,
+                         uint32_t (*GetStreamDepth)(void));
```
Comment (inliniac) [inline src/app-layer-parser.h]: formatting still looks off https://github.com/inliniac/suricata/pull/2307#pullrequestreview-2092317

File: doc/userguide/configuration/suricata-yaml.rst (hunk 1 above)
Comment (regit) [inline doc/userguide/configuration/suricata-yaml.rst]: ack

File: src/app-layer-modbus.c (hunk 2 above)
Comment (regit) [inline src/app-layer-modbus.c]: ack

File: src/app-layer-modbus.c (hunk 3 above)
Comment (regit) [inline src/app-layer-modbus.c]: ack

File: src/app-layer-parser.h (hunk 4 above)
Comment (regit) [inline src/app-layer-parser.h]: ack

**Pull request closed without merging**